### PR TITLE
feat(crypt): remove crypt keyfile before entering emergency shell

### PR DESCRIPTION
**Pull request by tblume**
RFC: Remove grub created keyfiles before entering emergency shell.
Since it isn't possible to distinguish whether a keyfile is grub created, do it for all key files, but add an option to suppress this behaviour.

## Checklist
- [X] I have tested it locally
- [X] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

Fixes #

### CODE REVIEW

File: modules.d/98dracut-systemd/module-setup.sh [hunk 1]
```diff
@@ -53,4 +53,37 @@ install() {
     inst_simple "$moddir/dracut-tmpfiles.conf" "$tmpfilesdir/dracut-tmpfiles.conf"
 
     inst_multiple sulogin
+
+    if [ -f "$dracutsysrootdir"/etc/crypttab ]; then
+        remove_cryptkeyfiles
```
Comment (aafeijoo-suse) [inline modules.d/98dracut-systemd/module-setup.sh:58]: IIRC the problem is key files generated by GRUB at boot time do not exist when dracut builds the initrd, so maybe this logic should be implemented in an emergency hook script (`lib/dracut/hooks/emergency`).

File: dracut.sh [hunk 2]
```diff
@@ -822,6 +825,7 @@ while :; do
             ;;
         --keep) keep="yes" ;;
         --printsize) printsize="yes" ;;
+        --keep-cryptkeyfiles)        keep_cryptkeyfiles="yes";;
```
Comment (aafeijoo-suse) [inline dracut.sh:824]: You need to export `keep_cryptkeyfiles` here:

https://github.com/openSUSE/dracut/blob/67527a99f1c5cad482d7bb78e68b5437a44b9027/dracut.sh#L1920-L1939

File: modules.d/98dracut-systemd/module-setup.sh (hunk 1 above)
Comment (tblume) [reply modules.d/98dracut-systemd/module-setup.sh:58]: The grub generated keyfiles have fixed path and names, so I can predict them.
But modifying the dracut-emergency.sh script when creating the initrd is probably not conform with the usual implementation methods.
I will switch to a hook.

File: dracut.sh (hunk 2 above)
Comment (tblume) [reply dracut.sh:824]: Ok, thanks!

File: modules.d/98dracut-systemd/module-setup.sh (hunk 1 above)
Comment (aafeijoo-suse) [reply modules.d/98dracut-systemd/module-setup.sh:58]: Although you can predict the name of the key file, I think GRUB appends the new entry to /etc/crypttab at boot time, but only to the /etc/crypttab file contained in the initrd, not to the system /etc/crypttab, so when dracut builds an initrd, that entry should not be there... but I think we should ask Michael to confirm it :)

Comment (aafeijoo-suse) [reply modules.d/98dracut-systemd/module-setup.sh:58]: > But modifying the dracut-emergency.sh script when creating the initrd is probably not conform with the usual implementation methods. I will switch to a hook.

I kept thinking about this and I think I expressed myself badly, just to clarify, the code on the `install` method would be like:

```bash
if [ -z "$keep_cryptkeyfiles" ]; then
    inst_hook emergency 000 "$moddir/remove_cryptkeyfiles.sh"
fi
```

Then, a new `remove_cryptkeyfiles.sh` script would contain the logic of your `remove_cryptkeyfiles` function (you only need to replace `-z "$keep_cryptkeyfiles"` with `-f /etc/crypttab`).

Comment (tblume) [reply modules.d/98dracut-systemd/module-setup.sh:58]: > Although you can predict the name of the key file, I think GRUB appends the new entry to /etc/crypttab at boot time, but only to the /etc/crypttab file contained in the initrd, not to the system /etc/crypttab, so when dracut builds an initrd, that entry should not be there... but I think we should ask Michael to confirm it :)

Hm, I was assuming that there can't  be case where the name of the crypt device is not listet in /etc/crypttab of the running system.
I derived this from the code in /etc/grub.d/05_crypttab.
But indeed, I might have misinterpreted the functionality.
@WenhuaChang does grub still write a keyfile for encrypted system root, even if the crypttab in the running system doesn't container an entry for this device?

Comment (aafeijoo-suse) [reply modules.d/98dracut-systemd/module-setup.sh:58]: > Hm, I was assuming that there can't be case where the name of the crypt device is not listet in /etc/crypttab of the running system.

Ahh, thanks for the reference, I was missing this part.
Then your code is correct, sorry for the noise.

**Review by aafeijoo-suse — APPROVED** [review]
Looks good.
@tblume, are there any plans to send this upstream in the near future, or can we merge it here until then?

**Comment by tblume** [comment]
> Looks good. @tblume, are there any plans to send this upstream in the near future, or can we merge it here until then?

Yes, will do that, so you can take it from upstream instead of merging this one.
But I want do do some more testing before.

**Comment by aafeijoo-suse** [comment]
Hey @tblume what have you done here? :)

**Comment by tblume** [comment]
> Hey @tblume what have you done here? :)

Outch sorry, I did a wrong push for the upstream submission.
However, I guess this one is finished.
I did:

https://github.com/dracutdevs/dracut/pull/2206

instead.
Will close this one.

**Pull request closed without merging**